kotlinx-coroutines-core/common/src/channels/ArrayChannel.kt

```diff
@@ -254,17 +259,23 @@ internal open class ArrayChannel<E>(
     // Note: this function is invoked when channel is already closed
     override fun onCancelIdempotent(wasClosed: Boolean) {
         // clear buffer first, but do not wait for it in helpers
-        if (wasClosed) {
-            lock.withLock {
-                repeat(size.value) {
-                    buffer[head] = 0
-                    head = (head + 1) % buffer.size
+        val onElementCancel = onElementCancel
+        var elementCancelException: ElementCancelException? = null // first cancel exception, others suppressed
+        lock.withLock {
+            repeat(size.value) {
+                val value = buffer[head]
+                if (value !== EMPTY && onElementCancel != null) {
```

Comment:
[nit] `onElementCancel != null && value !== EMPTY` is a bit faster and reflects the intent better

Reply:
Fixed